src/test/runner.ts

```diff
@@ -230,6 +230,43 @@ export class Runner {
 
       await Promise.all(Array.from(outputDirs).map(outputDir => removeFolderAsync(outputDir).catch(e => {})));
 
+      let testGroups = createTestGroups(rootSuite);
+
+      const shard = config.shard;
+      if (shard) {
```

Comment:
maybe an own function like we have for filterByFocusedLine and filterOnly since the logic from 237 to 260 is pretty specific for shards?

```suggestion
      if (shard)
        testGroups = filterGroupByShard(shard, rootSuite)
```